### PR TITLE
Update Bloop Java Home to that of Metals

### PR DESCRIPTION
**Pull request by ghost**
If Metals Java Home gets updated, the Bloop java home gets also updated with that of the metals, through the Bloop global json file.

### CODE REVIEW

File: metals/src/main/scala/scala/meta/internal/metals/UserConfiguration.scala [hunk 1]
```diff
@@ -39,6 +39,7 @@ case class UserConfiguration(
     bloopSbtAlreadyInstalled: Boolean = false,
     bloopVersion: Option[String] = None,
     bloopJvmProperties: Option[List[String]] = None,
+    bloopJavaHome: Option[String] = None,
```
Comment (tgodzik) [inline metals/src/main/scala/scala/meta/internal/metals/UserConfiguration.scala:42]: I would just use the same java Home as the Metals one it will be a bit confusing for the users otherwise.

**Review by tgodzik — CHANGES_REQUESTED** [review]

File: metals/src/main/scala/scala/meta/internal/metals/UserConfiguration.scala (hunk 1 above)
Comment (zmerr) [reply metals/src/main/scala/scala/meta/internal/metals/UserConfiguration.scala:42]: Ok, so I would remove it!

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala [hunk 2]
```diff
@@ -276,76 +273,89 @@ final class BloopServers(
...
-                )
-              case Success(_) => Future.successful()
-            }
-          }
-        } else Future.successful(())
 
-    result.getOrElse(Future.successful(()))
+        requestedBloopJvmProperties = maybeRequestedBloopJvmProperties
+          .getOrElse(List.empty)
+        if maybeRequestedBloopJvmProperties != maybeRunningBloopJvmProperties ||
```
Comment (tgodzik) [inline metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:289]: Could we rework `maybeRunningBloopJvmProperties` to read from the `bloop.json` file so that even if we have `None` then we will not show the message about restarting if it's not needed? Currently this shows up every time we start the server.

The same with `maybeRunningMetalsJavaHome`, it will be None in the default value and updated by the client each time. We should check if it's already set as Java home. I am thinking we might need an additional workflow for this actually.

We could set the `bloop.json` at the start of the server before starting Bloop (you can take a look at the BloopServers class) so that in most cases users don't need to do anything and not even ask about anything unless there is a `bloop.json` file already with a different home. In case bloop.json exists with a different home we can show a message to change the javaHome to avoid issues.

**Review by tgodzik — CHANGES_REQUESTED** [review]

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala (hunk 2 above)
Comment (zmerr) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala]: @tgodzik not sure if I understood what should be done for javaHome! But with the changes I just pushed, Metals does not propose to restart Bloop anymore, if the user requested setting matches the contents of the global json file! You can take a look!

**Review by tgodzik — CHANGES_REQUESTED** [review]

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala [hunk 3]
```diff
@@ -276,76 +294,95 @@ final class BloopServers(
   def ensureDesiredJvmSettings(
       maybeRequestedBloopJvmProperties: Option[List[String]],
       maybeRunningBloopJvmProperties: Option[List[String]],
-      maybeJavaHome: Option[String],
+      maybeRequestedMetalsJavaHome: Option[String],
+      maybeRunningMetalsJavaHome: Option[String],
```
[the remark this answers is not in the document]
Comment (zmerr) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:298]: Well, the way it works is this: it checks the running version from memory, then if it is None, reads from the json file! By removing them we would have to read from the json file each time by default! Is this the functionality we are looking for?

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala [hunk 4]
```diff
@@ -241,6 +241,27 @@ final class BloopServers(
     }.toOption
   }
 
+  private def maybeLoadBloopGlobalJsonFile(
+      bloopGlobalJsonFilePath: AbsolutePath
+  ): (Option[String], Option[List[String]]) = {
+    val maybeData = util.Try {
```
[the remark this answers is not in the document]
Comment (zmerr) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:247]: I remember last time I tried that, there were syntax highlighting issues! I would try again!

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:298]: > Well, the way it works is this: it checks the running version from memory, then if it is None, reads from the json file

I don't see where that happens, this condition:
```
 (maybeRequestedBloopJvmProperties != maybeRunningBloopJvmProperties
          && maybeRequestedBloopJvmProperties != maybeBloopGlobalJsonJvmProperties)
```
will be true at the start when both running java home and properties are None and the requested ones are Some. So we will ask to restart even if Bloop is running properly already.

It also seems we need to change both at the same time to get the pop up about restarting Bloop. It might be a bit less complex to use the global setting only instead of the one in the user configuration.

[the remark this answers is not in the document]
Comment (zmerr) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:298]: At the start, `maybeRequestedBloopJvmProperties != maybeRunningBloopJvmProperties` will be true! Then `maybeRequestedBloopJvmProperties != maybeBloopGlobalJsonJvmProperties` will be false! This means that 
`(maybeRequestedBloopJvmProperties != maybeRunningBloopJvmProperties
          && maybeRequestedBloopJvmProperties != maybeBloopGlobalJsonJvmProperties)` will be false! 
Therefor, at the start we are not going to ask to restart Bloop!

[the remark this answers is not in the document]
Comment (zmerr) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:298]: > It also seems we need to change both at the same time to get the pop up about restarting Bloop
  No, how did you deduct that?
according to this condition:

 if (maybeRequestedBloopJvmProperties != maybeRunningBloopJvmProperties
           && maybeRequestedBloopJvmProperties != maybeBloopGlobalJsonJvmProperties) ||
           (maybeRequestedMetalsJavaHome != maybeRunningMetalsJavaHome
             && maybeRequestedMetalsJavaHome != maybeBloopGlobalJsonJavaHome)
changing just one of either `bloopJvmProperties` or `javaHome` triggers the update! Because they are connected by `||`

[the remark this answers is not in the document]
Comment (zmerr) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:298]: > Well, the way it works is this: it checks the running version from memory, then if it is None, reads from the json file!

I guess the problem is that we are always reading from the bloop json file anyway! So I think I should just remove the `running` method arguments or change the semantics to the quoted logic?

[the remark this answers is not in the document]
Comment (zmerr) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:298]: I can right it as just 
```
 if maybeRequestedBloopJvmProperties != maybeBloopGlobalJsonJvmProperties ||
            maybeRequestedMetalsJavaHome != maybeBloopGlobalJsonJavaHome
```
for simplicity!

[the remark this answers is not in the document]
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:298]: Ach, you are right I misread the conditions, which is a perfect example of it being a bit complicated :sweat: 

Maybe we don't need the running since each time we change to requested that gets saved to global config, which should be a good source of truth.

[the remark this answers is not in the document]
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:298]: Yep, I think it should be much simpler this way.

[the remark this answers is not in the document]
Comment (zmerr) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:298]: Ok, no problem!

**Comment by tgodzik** [comment]
If anyone else could also take a look and double check the approach here I would be super thankful!

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala [hunk 5]
```diff
@@ -275,77 +293,92 @@ final class BloopServers(
...
+          maybeRequestedMetalsJavaHome != maybeBloopGlobalJsonJavaHome
+      } yield updateBloopJvmProperties(
+        requestedBloopJvmProperties,
+        bloopGlobalJsonFilePath,
+        bloopCreatedByMetalsFilePath,
+        maybeRequestedMetalsJavaHome,
+        reconnect
+      )
+    result.getOrElse {
+      Future.successful()
```
Comment (kpodsiad) [inline metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:320]:
```suggestion
      Future.unit
```

**Review by kpodsiad — COMMENTED** [review]
Could you remind me what are the benefits from coupling Metals' java and Bloop's java?

**Review by dos65 — COMMENTED** [review]
I don't mind of having this approach.

The only I'm worried is that javaHome changes silently lead to a notification `bloop properties was changed: Restart/Now now` that might be not clear for a user.

**Comment by tgodzik** [comment]
> I don't mind of having this approach.
> 
> The only I'm worried is that javaHome changes silently lead to a notification `bloop properties was changed: Restart/Now now` that might be not clear for a user.

@zmerr could we make sure that when java Home changes we get a different messages than when properties change?

**Comment by zmerr** [comment]
> > I don't mind of having this approach.
> > The only I'm worried is that javaHome changes silently lead to a notification `bloop properties was changed: Restart/Now now` that might be not clear for a user.
> 
> @zmerr could we make sure that when java Home changes we get a different messages than when properties change?

Sure! so as to avoid surprises for the user!

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala (hunk 5 above)
Comment (zmerr) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:320]: All applied!

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala [hunk 6]
```diff
@@ -241,6 +243,27 @@ final class BloopServers(
...
+      ujson.read(jsonString)
+    }
+    val javaHome = maybeData.flatMap { data =>
+      Try(data("javaHome").str)
+    }.toOption
+    val javaOptions = maybeData.flatMap { data =>
+      Try {
+        data("javaOptions").arr.toList.map(_.str)
+      }
+    }
```
Comment (tgodzik) [inline metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:263]:
```suggestion
     val javaHome = for {
        obj <- maybeData.objOpt
        strValue <- obj.get("javaHome")
        str <- strValue.strOpt
      } yield str
    val javaOptions = for {
      obj <- maybeData.objOpt
      value <- obj.get("javaOptions")
      arr <- value.arrOpt
      str = arr.flatMap(_.strOpt)
   } yield str
```

No need for the Try this way and it's better to handle things explicitely.

**Review by tgodzik — CHANGES_REQUESTED** [review]
Looks almost ready to merge. Just a few minor comments and you can change it from the draft PR to a normal one.

File: metals/src/main/scala/scala/meta/internal/metals/Messages.scala [hunk 7]
```diff
@@ -299,14 +323,20 @@ object Messages {
...
-    def params(): ShowMessageRequestParams = {
+
+    def params(
+        bloopJsonUpdateCause: BloopJsonUpdateCause
+    ): ShowMessageRequestParams = {
       val params = new ShowMessageRequestParams()
       params.setMessage(
-        s"""|You have previously modified the JVM settings of Bloop in its global json file,
+        s"""|Setting $bloopJsonUpdateCause will result in updating Bloop's global Json file, by Metals.
+            |But, you have previously modified this file with other values!
```
[the remark this answers is not in the document]
Comment (zmerr) [reply metals/src/main/scala/scala/meta/internal/metals/Messages.scala:339]: Are there any cases in which Bloop global json file’s modification could be automated, other than Metals?

[the remark this answers is not in the document]
Comment (zmerr) [reply metals/src/main/scala/scala/meta/internal/metals/Messages.scala]: applied. Thanks!

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala (hunk 6 above)
Comment (zmerr) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala]: applied. Thanks!

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala [hunk 8]
```diff
@@ -241,6 +243,27 @@ final class BloopServers(
...
+  private def maybeLoadBloopGlobalJsonFile(
+      bloopGlobalJsonFilePath: AbsolutePath
+  ): (Option[String], Option[List[String]]) = {
+    val maybeData = Try {
+      val source = scala.io.Source.fromFile(bloopGlobalJsonFilePath.toURI)
+      val jsonString =
+        try source.mkString
+        finally source.close()
+      ujson.read(jsonString)
+    }
```
[the remark this answers is not in the document]
Comment (zmerr) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala]: applied. Thanks!

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala [hunk 9]
```diff
@@ -275,77 +295,101 @@ final class BloopServers(
    */
   def ensureDesiredJvmSettings(
       maybeRequestedBloopJvmProperties: Option[List[String]],
```
Comment (tgodzik) [inline metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:297]: I think the issue with test is coming from the fact, that we are sending `None` here and we already have something set for CI. None should be treated as `use whatever is set currently` while `Some(Nil)` would be `remove all the options`

Comment (zmerr) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:297]: Thank you for this. I think the guard 
```
bloopJsonUpdateCause <-
          if (
            maybeRequestedBloopJvmProperties != maybeBloopGlobalJsonJvmProperties
          ) Some(BloopJsonUpdateCause.JVM_OPTS)
          else if (maybeRequestedMetalsJavaHome != maybeBloopGlobalJsonJavaHome)
            Some(BloopJsonUpdateCause.JAVA_HOME)
          else None

```
was enforcing that! Still, as an extra test, I checked that the `maybeRequestedBloopJvmProperties` is never `None`
It can only be `List.empty` when the user has not set anything!

Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:303]: I think it would be enought to add and if:
`if maybeRequestedBloopJvmProperties.nonEmpty`
somewhere in the for comprehension

Comment (zmerr) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:303]: Then what about the cases when it’s empty, while javaHome is set?

Comment (zmerr) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:303]: I think it might have to do with how the language client is called.

Comment (zmerr) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:303]: specifically in cases when the bloop global json is manually premodified. The errors do not show for the other case.

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala [hunk 10]
```diff
@@ -275,77 +301,100 @@ final class BloopServers(
...
-              case Failure(exception) =>
-                languageClient.showMessage(
-                  MessageType.Error,
-                  exception.getMessage
-                )
-              case Success(_) => Future.successful()
-            }
-          }
-        } else Future.successful(())
+            maybeRequestedBloopJvmProperties != maybeBloopGlobalJsonJvmProperties
```
Comment (tgodzik) [inline metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:319]:
```suggestion
     maybeRequestedBloopJvmProperties.nonEmpty &&   maybeRequestedBloopJvmProperties != maybeBloopGlobalJsonJvmProperties
```

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala [hunk 11]
```diff
@@ -120,29 +121,35 @@ final class BloopServers(
...
       bloopGlobalJsonFilePath: AbsolutePath,
       bloopCreatedByMetalsFilePath: AbsolutePath,
-      requestedBloopJvmProperties: List[String],
+      maybeRequestedBloopJvmProperties: Option[List[String]],
       maybeJavaHome: Option[String]
   ): Try[Unit] = Try {
 
-    val javaOptionsString =
-      s"\"javaOptions\": [${requestedBloopJvmProperties.map(property => s"\"$property\"").mkString(", ")}]"
+    val javaOptionsString = maybeRequestedBloopJvmProperties
```
Comment (tgodzik) [inline metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:138]: Btw. if `maybeRequestedBloopJvmProperties` is empty we might want to use the existing options that are in the file.

**Review by tgodzik — APPROVED** [review]
LGTM!
